### PR TITLE
ci: fix check for unused Rust dependencies

### PR DESCRIPTION
**Pull request by teskje**
This PR fixes the "Unused dependencies" nightly job, which is AFAICT the only one still consistently failing.

To make it work again I had to:
* Update cargo-udeps, to make it work with the new workspace inheritance syntax. I've updated the other cargo dependencies too while I was at it anyway, except for sccache whose current version doesn't compile with our feature config (https://github.com/mozilla/sccache/issues/1170).
* Force crosstool to use zlib 1.2.13, to work around https://github.com/crosstool-ng/crosstool-ng/issues/1832.
* Remove unused dependencies.

### Motivation

  * This PR fixes a previously unreported bug.

The "Unused dependencies" job is failing since forever (?).

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by teskje** [comment]
Successful run: https://buildkite.com/materialize/nightlies/builds/1300#01841ebd-1f62-4da2-b73a-fafc698f5603

**Review by philip-stoev — APPROVED** [review]
As this is a type of linter, ideally it should run on the per-push CI. However, it takes 6 minutes so that is too much carbon to burn for this. So Nightly appears to be the logical place right now, but can you please remove the `soft_fail: true`  from ci/nightly/pipeline-template.yaml so that it does not start failing silently again.

File: ci/builder/Dockerfile [hunk 1]
```diff
@@ -60,6 +60,8 @@ COPY crosstool-$ARCH_GCC.defconfig ./
 
 RUN DEFCONFIG=crosstool-$ARCH_GCC.defconfig ct-ng defconfig \
     && rm crosstool-$ARCH_GCC.defconfig \
+    # https://github.com/crosstool-ng/crosstool-ng/issues/1832 \
+    && sed -i 's/CT_ZLIB_VERSION="1.2.12"/CT_ZLIB_VERSION="1.2.13"/g' .config \
```
[the remark this answers is not in the document]
Comment (teskje) [reply ci/builder/Dockerfile:64]: We can add a comment to where we download crosstools, to remove the workaround when we update to a new version.

**Comment by teskje** [comment]
> So Nightly appears to be the logical place right now, but can you please remove the soft_fail: true from ci/nightly/pipeline-template.yaml so that it does not start failing silently again.

Apparently the reason we marked it soft-fail in the first place is this unresolved cargo-udeps bug: https://github.com/est31/cargo-udeps/issues/103. It's not resolved, but the last reported occurrence also was a while ago, so idk, might be worth a try?

**Comment by benesch** [comment]
The failure was always relatively rare! Like a low single digit percentage probability of failure. That was too high for every PR build, but I’d be much more willing to gamble with those odds on the nightly build. Can always soft fail it again if it becomes disruptive. 

Thanks for fixing this!

**Pull request merged**